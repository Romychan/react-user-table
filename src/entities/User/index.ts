export {
  useFetchUsers,
  useFetchUser,
  useCreateUser,
  useDeleteUser,
  useUpdateUser,
} from './api/queries';
export { userHandlers } from './api/msw/userHandlers';
export type { UsersResponse, UserDTO } from './api/types';
export { MOCK_USER } from './api/__mocks__/mocks';

export { UserContextProvider, UserContext } from './model/context';
export { useUserStore } from './model/hooks';
export {
  TABLE_USERS_COLUMNS,
  DEFAULT_USERS_PARAMS,
  FORM_USERS_FIELDS,
} from './model/constants';
export type { User, UserQueryParams, UserForm } from './model/types';
export { UserFormSchema } from './model/schemas';

export { UserAvatar } from './ui/UserAvatar/UserAvatar';
