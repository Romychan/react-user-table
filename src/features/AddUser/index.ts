export { AddUser } from './ui/AddUser';
