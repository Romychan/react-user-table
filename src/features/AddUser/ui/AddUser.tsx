import { Drawer } from '~/shared/ui/Drawer';
import { useToggle } from '~/shared/lib/hooks';
import { Button } from '~/shared/ui/Button';
import { Icon } from '~/shared/ui/Icon';

import { AddForm } from './AddForm/AddForm';
import styles from './AddUser.module.scss';

/** This component is used to display the form for adding a new user */
export const AddUser = () => {
  const { isToggled: isVisible, onToggle: toggleVisible } = useToggle();

  return (
    <div className={styles.container}>
      <Button
        size="sm"
        onClick={toggleVisible}
        className={styles.button}
        data-testid="drawer-add-user"
      >
        <Icon name="plus" size={10} />
        Add user
      </Button>

      <Drawer title="New user" isVisible={isVisible} onClose={toggleVisible}>
        {(onClose) => <AddForm onComplete={onClose} />}
      </Drawer>
    </div>
  );
};
