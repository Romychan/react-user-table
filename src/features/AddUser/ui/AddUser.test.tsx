import {
  fireEvent,
  screen,
  render,
  RenderResult,
} from '@testing-library/react';
import { Mock, vi } from 'vitest';

import * as UserEntity from '~/entities/User';

import { AddUser } from './AddUser';

describe('AddUser', () => {
  let mockMutate: Mock<never, never>;
  let container: RenderResult;

  beforeEach(() => {
    mockMutate = vi.fn();

    vi.spyOn(UserEntity, 'useCreateUser').mockImplementation(
      vi.fn().mockReturnValue({
        mutate: mockMutate,
      }),
    );

    container = render(<AddUser />);
  });

  it('should render correctly', () => {
    expect(screen.getByTestId('drawer-add-user')).toBeInTheDocument();
    expect(screen.queryByTestId('drawer')).not.toBeInTheDocument();
  });

  it('should open the drawer with the form correctly', () => {
    const buttonElement = screen.getByTestId('drawer-add-user');

    fireEvent.click(buttonElement);

    const drawerElement = screen.getByTestId('drawer');

    expect(drawerElement).toBeInTheDocument();
  });

  it('should close the drawer with the form correctly', () => {
    const buttonElement = screen.getByTestId('drawer-add-user');

    fireEvent.click(buttonElement);

    const drawerElement = screen.getByTestId('drawer');
    const cancelButtonElement = screen.getByTestId('cancel-add-user');
    const drawerBackdropElement = screen.getByTestId('drawer-backdrop');

    expect(drawerElement).toBeInTheDocument();

    fireEvent.click(cancelButtonElement);
    fireEvent.animationEnd(drawerBackdropElement);

    expect(screen.queryByTestId('drawer')).not.toBeInTheDocument();
  });

  it('should match snapshot', () => {
    expect(container).toMatchSnapshot();
  });
});
