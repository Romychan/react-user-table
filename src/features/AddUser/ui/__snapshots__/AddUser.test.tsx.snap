// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AddUser > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="_container_fcd5d4"
      >
        <button
          class="_button_0302f3 _button_fcd5d4 _primary_0302f3 _sm_0302f3"
          data-testid="drawer-add-user"
        >
          <svg
            class="icon icon-plus"
            data-testid="icon"
            height="10"
            width="10"
          >
            <use
              xlink:href="/images/icons.svg#plus"
            />
          </svg>
          Add user
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="_container_fcd5d4"
    >
      <button
        class="_button_0302f3 _button_fcd5d4 _primary_0302f3 _sm_0302f3"
        data-testid="drawer-add-user"
      >
        <svg
          class="icon icon-plus"
          data-testid="icon"
          height="10"
          width="10"
        >
          <use
            xlink:href="/images/icons.svg#plus"
          />
        </svg>
        Add user
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
